<?xml version="1.0" ?>
<robot name="nemo4b">
   <link name="pelvis">
      <inertial>
         <origin xyz="0.0 0.0 0.2" rpy="0.0 0.0 0.0"/>
         <mass value="5.11537484"/>
         <inertia ixx="0.03" iyy="0.03" izz="0.03" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="pelvis">
         <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/pelvis.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <link name="l_hip_pitch">
      <inertial>
      <!-- pos="-0.01821923 0 -3.205e-05", mass="1.8670971"
           fullinertia="0.00265583 0.00363357 0.00358009 -2e-07 -1.09e-06 0" -->
      <origin xyz="-0.01821923 0.0 -0.00003205" rpy="0 0 0"/>
      <mass value="1.8670971"/>
      <inertia
        ixx="0.00265583" ixy="-0.0000002"  ixz="-0.00000109"
        iyy="0.00363357" iyz="0.0"
        izz="0.00358009"/>
    </inertial>
      <visual name="l_hip_pitch">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_hip_pitch.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="l_hip_pitch">
      <parent link="pelvis"/>
      <child link="l_hip_pitch"/>
      <origin xyz="0.0 0.0965 -0.112" rpy="0 0 0"/>
      <axis xyz="0.0 1.0 0.0"/>
      <limit lower="-1.3" upper="1.3" effort="47.4" velocity="32"/>
   </joint>
   <link name="l_hip_roll">
      <inertial>
         <origin xyz="-0.01740501 1.764e-05 -0.05845559" rpy="0.0004863223254489439 0.19610377555273417 0.0010392456801208367"/>
         <mass value="2.00459959"/>
         <inertia ixx="0.006611571483374219" iyy="0.006331359985392611" izz="0.0028985185312331704" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_hip_roll">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_hip_roll.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="l_hip_roll">
      <parent link="l_hip_pitch"/>
      <child link="l_hip_roll"/>
      <origin xyz="-0.07875 0.0 0.0" rpy="0 0 0"/>
      <axis xyz="1.0 0.0 0.0"/>
      <limit lower="-0.4" upper="1.5" effort="23.7" velocity="32"/>
   </joint>
   <link name="l_hip_yaw">
      <inertial>
         <origin xyz="3.205e-05 -0.0 -0.06053076999999999" rpy="0.001179318771453497 0.0002045579945115783 1.5707921576398078"/>
         <mass value="1.8670971"/>
         <inertia ixx="0.003633570040911599" iyy="0.0035800912854579733" izz="0.0026558286736304273" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_hip_yaw">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_hip_yaw.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="l_hip_yaw">
      <parent link="l_hip_roll"/>
      <child link="l_hip_yaw"/>
      <origin xyz="0.00715 0.0 -0.1263" rpy="0 0 0"/>
      <axis xyz="0.0 0.0 1.0"/>
      <limit lower="-1.0" upper="1.0" effort="23.7" velocity="32"/>
   </joint>
   <link name="l_knee">
      <inertial>
         <origin xyz="0.0 5e-05 -0.10289321" rpy="0.0 0.0 0.0"/>
         <mass value="0.58007416"/>
         <inertia ixx="0.0070466" iyy="0.00513742" izz="0.00225232" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_knee">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_knee.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="l_knee">
      <parent link="l_hip_yaw"/>
      <child link="l_knee"/>
      <origin xyz="0.0 0.0 -0.07875" rpy="0 0 0"/>
      <axis xyz="0.0 1.0 0.0"/>
      <limit lower="-2.0" upper="2.0" effort="47.4" velocity="32"/>
   </joint>
   <link name="l_foot_pitch">
      <inertial>
         <origin xyz="0.01821923 -0.0 -3.205e-05" rpy="1.571975646419184 4.169155001454428e-06 1.57100088478941"/>
         <mass value="1.8670971"/>
         <inertia ixx="0.003633570040911599" iyy="0.0035800912854579733" izz="0.0026558286736304273" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_foot_pitch">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_foot_pitch.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="l_foot_pitch">
      <parent link="l_knee"/>
      <child link="l_foot_pitch"/>
      <origin xyz="0.0 0.0 -0.206" rpy="0 0 0"/>
      <axis xyz="0.0 1.0 0.0"/>
      <limit lower="-2.0" upper="2.0" effort="47.4" velocity="32"/>
   </joint>
   <link name="l_foot_roll">
      <inertial>
         <origin xyz="0.00757956 1.139e-05 -0.03942379" rpy="2.3433878067998437 -0.0008879622578734647 1.569836156558178"/>
         <mass value="1.5526793"/>
         <inertia ixx="0.005416700784880941" iyy="0.004963623876621819" izz="0.0028013153384972476" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_foot_roll">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_foot_roll.stl"/>
         </geometry>
         <material name="white"/>
      </visual>

      <collision>
        <origin xyz="-0.0187 0 -0.07975" rpy="0 0 0"/>
        <geometry>
          <box size="0.1 0.0505 0.006"/>
        </geometry>
      </collision>
   </link>
   <joint type="revolute" name="l_foot_roll">
      <parent link="l_foot_pitch"/>
      <child link="l_foot_roll"/>
      <origin xyz="0.07875 0.0 0.0" rpy="0 0 0"/>
      <axis xyz="1.0 0.0 0.0"/>
      <limit lower="-1.0" upper="1.0" effort="23.7" velocity="32"/>
   </joint>
   <link name="r_hip_pitch">
      <inertial>
         <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
         <mass value="1.8670971"/>
         <inertia ixx="0.003633570040911599" iyy="0.0035800912854579733" izz="0.0026558286736304273" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="r_hip_pitch">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_hip_pitch.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="r_hip_pitch">
      <parent link="pelvis"/>
      <child link="r_hip_pitch"/>
      <origin xyz="0.0 -0.0965 -0.112" rpy="0.0 0.0 0.0"/>
      <axis xyz="0.0 1.0 0.0"/>
      <limit lower="-1.3" upper="1.3" effort="47.4" velocity="32"/>
   </joint>
   <link name="r_hip_roll">
      <inertial>
         <origin xyz="-0.01740501 1.764e-05 -0.05845559" rpy="0.0004863223254489439 0.19610377555273417 0.0010392456801208367"/>
         <mass value="2.00459959"/>
         <inertia ixx="0.006611571483374219" iyy="0.006331359985392611" izz="0.0028985185312331704" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="r_hip_roll">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_hip_roll.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="r_hip_roll">
      <parent link="r_hip_pitch"/>
      <child link="r_hip_roll"/>
      <origin xyz="-0.07875 0.0 0.0" rpy="0 0 0"/>
      <axis xyz="1.0 0.0 0.0"/>
      <limit lower="-1.5" upper="0.4" effort="23.7" velocity="32"/>
   </joint>
   <link name="r_hip_yaw">
      <inertial>
         <origin xyz="3.205e-05 -0.0 -0.06053076999999999" rpy="0.001179318771453497 0.0002045579945115783 1.5707921576398078"/>
         <mass value="1.8670971"/>
         <inertia ixx="0.003633570040911599" iyy="0.0035800912854579733" izz="0.0026558286736304273" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="r_hip_yaw">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_hip_yaw.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="r_hip_yaw">
      <parent link="r_hip_roll"/>
      <child link="r_hip_yaw"/>
      <origin xyz="0.00715 0.0 -0.1263" rpy="0 0 0"/>
      <axis xyz="0.0 0.0 1.0"/>
      <limit lower="-1.0" upper="1.0" effort="23.7" velocity="32"/>
   </joint>
   <link name="r_knee">
      <inertial>
         <origin xyz="0.0 5e-05 -0.10289321" rpy="0.0 0.0 0.0"/>
         <mass value="0.58007416"/>
         <inertia ixx="0.0070466" iyy="0.00513742" izz="0.00225232" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_knee">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_knee.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="r_knee">
      <parent link="r_hip_yaw"/>
      <child link="r_knee"/>
      <origin xyz="0.0 0.0 -0.07875" rpy="0 0 0"/>
      <axis xyz="0.0 1.0 0.0"/>
      <limit lower="-2.0" upper="2.0" effort="100" velocity="100"/>
   </joint>
   <link name="r_foot_pitch">
      <inertial>
         <origin xyz="0.01821923 -0.0 -3.205e-05" rpy="1.571975646419184 4.169155001454428e-06 1.57100088478941"/>
         <mass value="1.8670971"/>
         <inertia ixx="0.003633570040911599" iyy="0.0035800912854579733" izz="0.0026558286736304273" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_foot_pitch">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_foot_pitch.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="r_foot_pitch">
      <parent link="r_knee"/>
      <child link="r_foot_pitch"/>
      <origin xyz="0.0 0.0 -0.206" rpy="0 0 0"/>
      <axis xyz="0.0 1.0 0.0"/>
      <limit lower="-2.0" upper="2.0" effort="47.4" velocity="32"/>
   </joint>
   <link name="r_foot_roll">
      <inertial>
         <origin xyz="0.00757956 1.139e-05 -0.03942379" rpy="2.3433878067998437 -0.0008879622578734647 1.569836156558178"/>
         <mass value="1.5526793"/>
         <inertia ixx="0.005416700784880941" iyy="0.004963623876621819" izz="0.0028013153384972476" ixy="0" ixz="0" iyz="0"/>
      </inertial>
      <visual name="l_foot_roll">
         <geometry>
            <mesh filename="package://prairie_control/meshes/nemo4b/l_foot_roll.stl"/>
         </geometry>
         <material name="white"/>
      </visual>
   </link>
   <joint type="revolute" name="r_foot_roll">
      <parent link="r_foot_pitch"/>
      <child link="r_foot_roll"/>
      <origin xyz="0.07875 0.0 0.0" rpy="0 0 0"/>
      <axis xyz="1.0 0.0 0.0"/>
      <limit lower="-1.0" upper="1.0" effort="23.7" velocity="32"/>
   </joint>
   <material name="white">
      <color rgba="1 1 1 1"/>
   </material>

   <gazebo>
    <plugin name="gz::sim::systems::JointStatePublisher" filename="gz-sim-joint-state-publisher-system">
      <update_rate>2000</update_rate>
        <joint_name>l_hip_pitch</joint_name>
        <joint_name>l_hip_roll</joint_name>
        <joint_name>l_hip_yaw</joint_name>
        <joint_name>l_knee</joint_name>
        <joint_name>l_foot_pitch</joint_name>
        <joint_name>l_foot_roll</joint_name>
        <joint_name>r_hip_pitch</joint_name>
        <joint_name>r_hip_roll</joint_name>
        <joint_name>r_hip_yaw</joint_name>
        <joint_name>r_knee</joint_name>
        <joint_name>r_foot_pitch</joint_name>
        <joint_name>r_foot_roll</joint_name>
    </plugin>
  </gazebo>

  <gazebo>
    <plugin name="gz::sim::systems::JointTrajectoryController" filename="gz-sim-joint-trajectory-controller-system">
      <joint_name>l_hip_pitch</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>50</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>4</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>l_hip_roll</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>40</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>2</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>l_hip_yaw</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>40</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>2</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>l_knee</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>50</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>4</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>l_foot_pitch</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>50</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>4</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>l_foot_roll</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>40</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>2</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>r_hip_pitch</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>50</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>4</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>r_hip_roll</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>40</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>2</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>r_hip_yaw</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>40</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>2</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>r_knee</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>50</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>4</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>r_foot_pitch</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>50</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>4</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>

      <joint_name>r_foot_roll</joint_name>
      <use_header_start_time>true</use_header_start_time>
      <position_p_gain>40</position_p_gain>
      <position_d_gain>0</position_d_gain>
      <velocity_p_gain>2</velocity_p_gain>
      <velocity_d_gain>0</velocity_d_gain>
      <position_cmd_min>-40</position_cmd_min>
      <position_cmd_max>40</position_cmd_max>
      <velocity_cmd_min>-10</velocity_cmd_min>
      <velocity_cmd_max>10</velocity_cmd_max>
    </plugin>
    <plugin
       filename="gz-sim-pose-publisher-system"
       name="gz::sim::systems::PosePublisher">
       <publish_link_pose>true</publish_link_pose>
       <use_pose_vector_msg>true</use_pose_vector_msg>
    </plugin>
    <plugin
       filename="gz-sim-odometry-publisher-system"
       name="gz::sim::systems::OdometryPublisher">
       <robot_base_frame>h1/pelvis</robot_base_frame>
       <dimensions>3</dimensions>
    </plugin>

  </gazebo>
  <gazebo reference="pelvis">
    <sensor name="imu_sensor" type="imu">
       <always_on>1</always_on>
       <update_rate>1</update_rate>
       <visualize>true</visualize>
       <topic>imu</topic>
       <plugin
        filename="gz-sim-imu-system"
        name="gz::sim::systems::Imu">
        </plugin>
    </sensor>
  </gazebo>
</robot>
